README.md
# https://github.com/docker-library/mysql

## Maintained by: [the Docker Community and the MySQL Team](https://github.com/docker-library/mysql)

This is the Git repo of the [Docker "Official Image"](https://docs.docker.com/docker-hub/official_repos/) for [mysql](https://hub.docker.com/_/mysql/) (not to be confused with any official mysql image provided by mysql upstream). See [the Docker Hub page](https://hub.docker.com/_/mysql/) for the full readme on how to use this Docker image and for information regarding contributing and issues.

The [full description from Docker Hub](https://hub.docker.com/_/mysql/) is generated over in [docker-library/docs](https://github.com/docker-library/docs), specifically in [docker-library/docs/mysql](https://github.com/docker-library/docs/tree/master/mysql).

## See a change merged here that doesn't show up on Docker Hub yet?

Check [the "library/mysql" manifest file in the docker-library/official-images repo](https://github.com/docker-library/official-images/blob/master/library/mysql), especially [PRs with the "library/mysql" label on that repo](https://github.com/docker-library/official-images/labels/library%2Fmysql).

For more information about the official images process, see the [docker-library/official-images readme](https://github.com/docker-library/official-images/blob/master/README.md).

---

-	[Travis CI:  
	![build status badge](https://img.shields.io/travis/docker-library/mysql/master.svg)](https://travis-ci.org/docker-library/mysql/branches)
-	[Automated `update.sh`:  
	![build status badge](https://doi-janky.infosiftr.net/job/update.sh/job/mysql/badge/icon)](https://doi-janky.infosiftr.net/job/update.sh/job/mysql)

| Build | Status | Badges | (per-arch) |
|:-:|:-:|:-:|:-:|
| [`amd64`<br />![build status badge](https://doi-janky.infosiftr.net/job/multiarch/job/amd64/job/mysql/badge/icon)](https://doi-janky.infosiftr.net/job/multiarch/job/amd64/job/mysql) |

<!-- THIS FILE IS GENERATED BY https://github.com/docker-library/docs/blob/master/generate-repo-stub-readme.sh -->


changelog
Changes in 1.1.5 (2018-04-19)
 * Bumped server versions to 5.5.60, 5.6.40, 5.7.22
 * MySQL Server 8.0 updated to 8.0.11
 * Bumped Shell 8.0 version to 8.0.11

Changes in 1.1.4 (2018-01-24)
 * MySQL Server 8.0 updated to 8.0.4-rc
 
Changes in 1.1.3 (2018-01-15)
 * Bumped server versions to 5.5.59, 5.6.39, 5.7.21
 * Bumped Shell 1.0 version to 1.0.11

Changes in 1.1.2 (2017-10-16)
 * Bumped server versions to 5.5.58, 5.6.38, 5.7.20
 * Bumped Shell 8.0 version to 8.0.3

Changes in 1.1.1 (2017-09-22)
 * Disable binlog for all entrypoint client commands (#87305)
   This was previously only done only for the script to set the root password,
   but should be done for all client commands, or group replication will fail
 * Add flag for logging to console (#87458)
   Setting -e MYSQL_LOG_CONSOLE=true will make the server log to console, so the
   server log is available through Docker's log interface. If /etc/my.cnf is
   mounted from the host system, this has no effect. On 8.0, this is the default
   behavior.
 * Ensure both MYSQL_USER and MYSQL_PASSWORD are set (#86982)
   Before, setting only one would cause the option to be silently ignored.
 * Read password file first and throw error if it is empty (#87025)
   Setting MYSQL_ROOT_PASSWORD to an empty string would be rejected, but
   setting it to a file with no content would be accepted. Both are now rejected.
 * MYSQL_RANDOM_ROOT_PASSWORD and MYSQL_ONETIME_PASSWORD are now set by default. (#87515)
   A password option no longer needs to be set explicitly. If none are set, a
   random, expired password is set.
 * Make root@$MYSQL_ROOT_HOST identical to root@localhost (#86945)
   If the MYSQL_ROOT_HOST variable is set we add an additional root user for
   the given host. This user was missing the PROXY privilege. We add the
   missing privilege so the users are identical.
 * MySQL Server 8.0 updated to 8.0.3-rc
 * MySQL Shell for 5.7 updated to 1.0.10

Changes in 1.1.0 (2017-07-24)
 * Fixed broken MYSQL_ONETIME_PASSWORD support (#86523)
   The functionality wasn't correctly updated when the entrypoint script was
   changed to only create the root@localhost user by default
 * Changed control files to be generated from a common template
   The control files for each version will be generated by swapping out the few
   commands that differ between server versions
 * Made PACKAGE_URL an ARG entry, to allow build-time changing
   The URL variables can be changed by supplying new ones with --build-arg at
   build time
 * Added basic healthcheck script
   Once database init is complete a flag file will be touched. If this file
   exists and mysqladmin ping succeeds, the container is reported as healthy,
   meaning it is ready for use
 * Database init will now use configured socket setting instead of hardcoded
   Using a hardcoded socket could cause init scripts to fail in confusing ways
   if the user customized the socket setting
 * Added MySQL Shell to supported images
   For 5.7 and 8.0, MySQL Shell is installed together with the server.
   https://dev.mysql.com/doc/refman/5.7/en/mysql-shell.html

Known issues:
 * For existing databases created with older images/not with Docker, the
   healthcheck script will generate an "Access denied for healthchecker@localhost"
   message in the server log every time it runs. If connection throttling is
   enabled this could eventually cause the healthcheck to fail


gen_dockerfiles.sh
#!/bin/bash
# Copyright (c) 2017, 2018 Oracle and/or its affiliates. All rights reserved.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; version 2 of the License.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA

set -e

function get_full_filename() {
        FILEPATH=$1
        PACKAGE_STRING=$2
        FILENAME=$(curl -s $FILEPATH/ | grep $PACKAGE_STRING | sed -e 's/.*href=\"//i' -e 's/\".*//')
        if [ -z "$FILENAME" ]; then
            echo &< "Unable to locate package for $PACKAGE_STRING. Aborting"
            exit 1
        fi
	COUNT=$(echo $FILENAME | tr " " "\n" | wc -l)
        if [ $COUNT -gt 1 ]; then
            echo &<2 "Found multiple file names for package $PACKAGE_STRING. Aborting"
            exit 1
        fi
	echo $FILENAME
}

# This script will simply use sed to replace placeholder variables in the
# files in template/ with version-specific variants.

. VERSION

if [ -z "$1" ]; then
  REPO=https://repo.mysql.com
else
  REPO=$1
fi

# 33060 is the default port for the mysqlx plugin, new to 5.7
declare -A PORTS
PORTS["5.5"]="3306"
PORTS["5.6"]="3306"
PORTS["5.7"]="3306 33060"
PORTS["8.0"]="3306 33060"

declare -A PASSWORDSET
PASSWORDSET["5.5"]="SET PASSWORD FOR 'root'@'localhost'=PASSWORD('\${MYSQL_ROOT_PASSWORD}');"
PASSWORDSET["5.6"]=${PASSWORDSET["5.5"]}
PASSWORDSET["5.7"]="ALTER USER 'root'@'localhost' IDENTIFIED BY '\${MYSQL_ROOT_PASSWORD}';"
PASSWORDSET["8.0"]=${PASSWORDSET["5.7"]}

declare -A DATABASE_INIT
DATABASE_INIT["5.5"]="mysql_install_db --user=mysql --datadir=\"\$DATADIR\" --rpm"
DATABASE_INIT["5.6"]="mysql_install_db --user=mysql --datadir=\"\$DATADIR\" --rpm --keep-my-cnf"
DATABASE_INIT["5.7"]="\"\$@\" --initialize-insecure"
DATABASE_INIT["8.0"]="\"\$@\" --initialize-insecure"

# 5.7+ has the --daemonize flag, which makes the process fork and then exit when
# the server is ready, removing the need for a fragile wait loop
declare -A INIT_STARTUP
INIT_STARTUP["5.5"]="\"\$@\" --skip-networking --socket=\"\$SOCKET\" \&"
INIT_STARTUP["5.6"]="\"\$@\" --skip-networking --socket=\"\$SOCKET\" \&"
INIT_STARTUP["5.7"]="\"\$@\" --daemonize --skip-networking --socket=\"\$SOCKET\""
INIT_STARTUP["8.0"]="\"\$@\" --daemonize --skip-networking --socket=\"\$SOCKET\""

declare -A STARTUP_WAIT
STARTUP_WAIT["5.5"]="\"yes\""
STARTUP_WAIT["5.6"]="\"yes\""
STARTUP_WAIT["5.7"]="\"\""
STARTUP_WAIT["8.0"]="\"\""

# The option to set a user as expired, (forcing a password change before
# any other action can be taken) was added in 5.6
declare -A EXPIRE_SUPPORT
EXPIRE_SUPPORT["5.5"]="\"\""
EXPIRE_SUPPORT["5.6"]="\"yes\""
EXPIRE_SUPPORT["5.7"]="\"yes\""
EXPIRE_SUPPORT["8.0"]="\"yes\""

# sed is for https://bugs.mysql.com/bug.php?id=20545
declare -A TZINFO_WORKAROUND
TZINFO_WORKAROUND["5.5"]="sed 's/Local time zone must be set--see zic manual page/FCTY/' | "
TZINFO_WORKAROUND["5.6"]="sed 's/Local time zone must be set--see zic manual page/FCTY/' | "
TZINFO_WORKAROUND["5.7"]=""
TZINFO_WORKAROUND["8.0"]=""

# Logging to console (stderr) makes server log available with the «docker logs command»
declare -A DEFAULT_LOG
DEFAULT_LOG["5.5"]=""
DEFAULT_LOG["5.6"]=""
DEFAULT_LOG["5.7"]=""
DEFAULT_LOG["8.0"]="console"


for VERSION in "${!MYSQL_SHELL_VERSIONS[@]}"
do
  # Dockerfiles
  MYSQL_SERVER_REPOPATH=yum/mysql-$VERSION-community/docker/x86_64
  MYSQL_CLUSTER_PACKAGE_URL=$REPO/$MYSQL_SERVER_REPOPATH/$(get_full_filename $REPO/$MYSQL_SERVER_REPOPATH mysql-community-server-minimal-${MYSQL_SERVER_VERSIONS[${VERSION}]})
  sed 's#%%MYSQL_SERVER_PACKAGE_URL%%#'"$MYSQL_CLUSTER_PACKAGE_URL"'#g' template/Dockerfile > tmpfile

  if [[ ! -z ${MYSQL_SHELL_VERSIONS[${VERSION}]} ]]; then
    MYSQL_SHELL_REPOPATH=yum/mysql-tools-community/el/7/x86_64
    MYSQL_SHELL_PACKAGE_URL=$REPO/$MYSQL_SHELL_REPOPATH/$(get_full_filename $REPO/$MYSQL_SHELL_REPOPATH mysql-shell-${MYSQL_SHELL_VERSIONS[${VERSION}]})
    sed -i 's#%%MYSQL_SHELL_PACKAGE_URL%%#'"$MYSQL_SHELL_PACKAGE_URL"'#g' tmpfile
  else
    sed -i 's#%%MYSQL_SHELL_PACKAGE_URL%%#'""'#g' tmpfile
  fi

  sed -i 's/%%PORTS%%/'"${PORTS[${VERSION}]}"'/g' tmpfile
  mv tmpfile ${VERSION}/Dockerfile

  # Entrypoint
  sed 's#%%PASSWORDSET%%#'"${PASSWORDSET[${VERSION}]}"'#g' template/docker-entrypoint.sh > tmpfile
  sed -i 's#%%DATABASE_INIT%%#'"${DATABASE_INIT[${VERSION}]}"'#g' tmpfile
  sed -i 's#%%EXPIRE_SUPPORT%%#'"${EXPIRE_SUPPORT[${VERSION}]}"'#g' tmpfile
  sed -i 's#%%SED_TZINFO%%#'"${TZINFO_WORKAROUND[${VERSION}]}"'#g' tmpfile
  sed -i 's#%%INIT_STARTUP%%#'"${INIT_STARTUP[${VERSION}]}"'#g' tmpfile
  sed -i 's#%%STARTUP_WAIT%%#'"${STARTUP_WAIT[${VERSION}]}"'#g' tmpfile
  sed -i 's#%%FULL_SERVER_VERSION%%#'"${FULL_SERVER_VERSIONS[${VERSION}]}"'#g' tmpfile
  sed -i 's#%%DEFAULT_LOG%%#'"${DEFAULT_LOG[${VERSION}]}"'#g' tmpfile
  mv tmpfile ${VERSION}/docker-entrypoint.sh
  chmod +x ${VERSION}/docker-entrypoint.sh

  # Healthcheck
  cp template/healthcheck.sh ${VERSION}/
  chmod +x ${VERSION}/healthcheck.sh
done
